src/plugins/dist_matrixops/all_gather.cpp

```diff
@@ -0,0 +1,383 @@
+                            "the same number of dimensions", name, codename));
+                }
+
+                std::array<std::size_t, PHYLANX_MAX_DIMENSIONS> dim =
+                    extract_numeric_value_dimensions(args[i], name, codename);
+
+                if (i != 0 && prevdim[0] != dim[0])
+                {
+                    HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                        "dist_matrixops::primitives::"
+                        "all_gather::detail::concatenate2d_axis1",
+                        util::generate_error_message(
+                            "all the input array dimensions except for"
+                            "the concatenation axis must match exactly",
+                            name, codename));
+                }
+
+                total_cols += dim[1];
+                prevdim = dim;
+            }
+
+            blaze::DynamicMatrix<T> result(prevdim[0], total_cols);
+
+            std::size_t step = 0;
+            for (auto&& arg : args)
+            {
+                auto&& val =
+                    execution_tree::extract_node_data<T>(std::move(arg));
+                std::size_t num_cols = val.dimension(1);
+                auto m = val.matrix();
+                for (std::size_t j = 0; j != num_cols; ++j)
+                {
+                    blaze::column(result, j + step) = blaze::column(m, j);
+                }
+                step += num_cols;
+            }
+
+            return execution_tree::primitive_argument_type{
+                ir::node_data<T>{std::move(result)}};
+        }
+
+        template <typename T>
+        execution_tree::primitive_argument_type concatenate2d(
+            execution_tree::primitive_arguments_type&& args, std::int64_t axis,
+            std::string const& name, std::string const& codename)
+        {
+            switch (axis)
+            {
+            case 0:
+                return detail::concatenate2d_axis0<T>(
+                    std::move(args), name, codename);
+            case 1:
+                return detail::concatenate2d_axis1<T>(
+                    std::move(args), name, codename);
+            default:
+                break;
+            }
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "dist_matrixops::primitives::"
+                "all_gather::detail::concatenate2d",
+                util::generate_error_message(
+                    "axis is out of bounds of dimension", name, codename));
+        }
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename T>
+    execution_tree::primitive_argument_type all_gather::all_gather2d(
+        ir::node_data<T>&& arr,
+        execution_tree::localities_information&& locs) const
+    {
+        using namespace execution_tree;
+        auto m = arr.matrix();
+        blaze::DynamicMatrix<T> res_value(m.rows(), m.columns());
+        res_value = m;
```

Comment:
can we write the above three lines as `blaze::DynamicMatrix<T> res_value = arr.matrix()` if we need to allocate memory?

Reply:
Yes. Will revise it.